### ISSUE TITLE
Privacy assessment .tsv files. 

### ISSUE THREAD

**Issue by EmiKib** [SIESTA-eu/wp15#38]
An addition to the .tsv scrambler could be to look into how much leakage there is in a scrambled dataset. 

Simple metrics could be full row leakage & partial row leakage with the option to set a value representing NaN. 

```python
def detect_row_leakage(df_original, df_scrambled, ignore_value= value):

    if df_original.shape != df_scrambled.shape:
        print("DataFrames do not have the same shape.")
        return

    total_rows = df_original.shape[0]  
    total_columns = df_original.shape[1]  
    partial_leakage_count = 0  
    full_leakage_count = 0  
    matching_cells_per_row = []  
    
    for idx, (row_orig, row_scram) in enumerate(zip(df_original.iterrows(), df_scrambled.iterrows())):
        row_orig = row_orig[1]
        row_scram = row_scram[1]
        valid_mask = (row_orig != ignore_value) & (row_scram != ignore_value)  
        
        # Count matches where both values are valid (not ignore_value)
        matches = (row_orig[valid_mask] == row_scram[valid_mask])
        match_count = matches.sum()
        matching_cells_per_row.append(match_count)
   
        # Check for full leakage
        if match_count == valid_mask.sum():
            full_leakage_count += 1
        # Check for partial leakage
        elif 0 < match_count < valid_mask.sum():
            partial_leakage_count += 1

    partial_leakage_percentage = (partial_leakage_count / total_rows) * 100 if total_rows > 0 else 0
    full_leakage_percentage = (full_leakage_count / total_rows) * 100 if total_rows > 0 else 0
    
    avg_matching_cells_per_row = np.mean(matching_cells_per_row) if total_rows > 0 else 0
    std_matching_cells_per_row = np.std(matching_cells_per_row) if total_rows > 0 else 0
    
    # Print the results
    print(f"Percentage of rows with partial leakage: {partial_leakage_percentage:.2f}%")
    print(f"Percentage of rows with full leakage: {full_leakage_percentage:.2f}%")
    print(f"Average number of matching cells per row: {avg_matching_cells_per_row:.2f} / {total_columns} fields per row")
    print(f"Standard deviation of matching cells per row: {std_matching_cells_per_row:.2f}")

```

In addition running a comparison between the Privacy Information Factor (implemented in the metaprivBIDS application) specifically the Field Information gain, would give an overview over how the original and srambled dataset assess the information gain for the individual columns. In that way the user has an overview over much information is retained in their scrambled dataset compared to their new. This would most often be almost the same as the permutation is row based, but it gives a quantifiable metric the user can refer back to. 

Lastly a  comparison of the scrambled and original datasets correlation matrix between columns could be an option to ensure that the dataset has been permuted to a satisfactory level. 

```python
def plot_correlation_matrices(df, df_scramble):
    # Detect categorical columns based on data type (object for strings, category)
    categorical_columns = df.select_dtypes(include=['object', 'category']).columns.tolist()

    for col in categorical_columns:
        df[col] = df[col].astype('category').cat.codes
        df_scramble[col] = df_scramble[col].astype('category').cat.codes

    corr_matrix1 = df.corr()
    corr_matrix2 = df_scramble.corr()

    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 6))  

    sns.heatmap(corr_matrix1, annot=False, cmap='coolwarm', linewidths=0.5, ax=ax1)
    ax1.set_title("Original Data Correlation")

    sns.heatmap(corr_matrix2, annot=False, cmap='coolwarm', linewidths=0.5, ax=ax2)
    ax2.set_title("Scrambled Data Correlation")

    plt.tight_layout()
    plt.show()

```

I hope this could be useful? If so, I could try to add it so it can run with the pipeline or as a separate pipeline.

**Comment by schoffelen** [comment]
for reference: https://github.com/CPernet/metaprivBIDS

**Comment by EmiKib** [comment]
@schoffelen does it fail at  "conda install graphviz pygraphviz" or at "pip install -e . " ?

**Comment by schoffelen** [comment]
> @schoffelen does it fail at "conda install graphviz pygraphviz" or at "pip install -e . " ?

Hi @EmiKib thanks for getting back about this. I must confess that I haven't tried 'conda install', I did pip install for the graphviz etc. 

I freshly cafeinated myself (and the compute cluster), and tried again just yet. With conda install I make it through the installation process. Thanks! PS: you may consider to add a line to the README.md to ```cd``` into the metaPrivBIDS repo (after git cloning it) before calling ```pip install -e .```

**Comment by EmiKib** [comment]
@schoffelen great that it works now. I will add that ```cd```  line now. 
If you have any suggestions in regards to anything in the app (design, buttons, functionality), albeit I am currently working on making it a bit more visually pleasing, please let me know. 
I am currently on holiday, but I will try to be fast at replying.

**Comment by schoffelen** [comment]
> @schoffelen great that it works now. I will add that `cd` line now. If you have any suggestions in regards to anything in the app (design, buttons, functionality), albeit I am currently working on making it a bit more visually pleasing, please let me know. I am currently on holiday, but I will try to be fast at replying.

Thanks, I will play around a bit. Don't bother replying during your holiday. Enjoy!

**Comment by CPernet** [comment]
- [x] Privacy assessment can be done using https://github.com/CPernet/metaprivBIDS --> command line tools via https://github.com/CPernet/metaprivBIDS/blob/main/metaprivBIDS/metaprivBIDS_core_logic.py 
- [x] Leakage can now be assessed using [DatLeak](https://github.com/SIESTA-eu/DatLeak)

The scrambler is there to create anonymous data, so there is no need to use metaprivBIDS tools to   
(1) assess the generated file as it won't tell anything about the anonymity of the generated data relative to the original one
(2) 'improve' anonymity with tools like such as adding noise or merging categories because those do not guarantee anonymity. The randomisation in the scrambler does exactly that job, and does it well and we can guaranty that by having 0% of full leakage and report/control partial leakage (which can only be more or less changed, depends on data)

**Comment by robertoostenveld** [comment]
@EmiKib, @CPernet and @marcelzwiers, is there any work still to be done on this? If not, please close this issue.

**Comment by robertoostenveld** [comment]
We discussed just now that some documentation needs to be added to https://github.com/SIESTA-eu/wp15/blob/main/doc/data_rights_holder.md

**Comment by robertoostenveld** [comment]
@CPernet will check the status of this with @marcelzwiers

**Comment by CPernet** [comment]
@Vincent9339 can you run [use case 2.1](https://github.com/SIESTA-eu/wp15/tree/main/usecase-2.1) including the scrambling, then add the leakage assessment using https://github.com/SIESTA-eu/DatLeak and document the code to use in the [README](https://github.com/SIESTA-eu/wp15/blob/main/usecase-2.1/README.md)

**Comment by Vincent9339** [comment]
Hi Cyril,

I had some technical difficulties with versioning, took me some time to fix
it.
Running this line *Rscript work/pipeline.R --inputdir input --outputdir
output  *gives me some statistic for some participants. In general r-based
container works fine. There's need for some manual work, but quite doable.

But running Scramble command takes forever and does not give any output.
Not even an error to figure out why.

I looked at the DataLeak and gives the result as expected in the README
file.

Do you want me to document DataLeak <https://github.com/SIESTA-eu/DatLeak>
code in use case 2.1
<https://github.com/SIESTA-eu/wp15/tree/main/usecase-2.1>?
Aside from that, will you also help me to get access to Confluence, so I
can read more about the project?

Overall a summary over what part of the SIESTA project should I focus, so I
continue reading the documents.

Thank you

**Comment by CPernet** [comment]
yes please propose changes in the readme, including the code change versioning issues you had to fix and how to use 
-- do branch and do a pull request, allowing to review the code change. 

(re confluence, Marcos emailed you yesterday with a link)

**Comment by marcelzwiers** [comment]
> But running Scramble command takes forever and does not give any output.
Not even an error to figure out why.

That's strange, but here scrambling tsv files is super quick. What command exactly did you use?
